### ISSUE TITLE
[RFC] Mutable HTTP request/response

### ISSUE THREAD

**Issue by chriskapp** [php-cds/php-cds#30]
there were many discussions about the immutable/mutable design of PSR-7 on the PGP-FIG mailing list. Eventually the objects are now immutable which was decided mostly by the designers of the PSR. With this issue I want to see how the community feels about this decision.

The bigest problem of immutability is that it is not easy possible to convert an application which is using a mutable response to an immutable implementation. Since the most popular HTTP implementation (Symfony/HttpFoundation) is mutable all applications which are using these objects probably rely on the mutability. I.e. with mutability it is easy to pass an HTTP response to a middleware or event listener which modifies the response. With immutability you also have to return the changed instance back to the application which is not always possible or requires a major change of the application design.

If there are also other people who would like to see such a mutable version I would post a first draft to this issue. Essentially this would mean to replace all with\* methods with set\* methods which are working on the object instead of returning a new object. If all people are happy with the current PSR-7 design it is also ok but I want to hear the feedback ;)

**Comment by CarwynNelson** [comment]
If i've read this correctly you are proposing an alternative to PSR-7. I'm not in favour of this for a few reasons:
- I personally feel that implementing PSR-7 as an "immutable interface" was the correct thing to do going forward, and whilst it might cause some growing pains it will ultimately result in a better situation for everyone.
- Competing with a PSR that is already in fairly wide use, and seems to have been accepted by the community quite well seems like a bad idea (This is different with PSR-6 as it seems to have gone down pretty poorly with the community, and I don't think it has seen wide adoption). Relevant xkcd: https://xkcd.com/927/
- Whilst frameworks like Symfony have built their `Request` and `Response` objects to be mutable I don't think there have been any issues in writing wrappers for PSR-7 which seem to work well, even though the underlying libraries are indeed mutable. eg: http://symfony.com/blog/psr-7-support-in-symfony-is-here

I understand where you are coming from with this suggestion but think our efforts are best served working on newer interfaces / recommendations, and trying to fix some of the issues with PSR's that might not've gone down well with the community.

Apologies, but this is a -1 from me.

**Comment by CarwynNelson** [comment]
@marcj @geggleto @xyNNN This RFC currently has more than 1/3 negative votes. If it is okay with the author of the RFC, and within the time frame allowed for an RFC to be closed, then could we please mark this topic as closed?

Cheers.

**Comment by xyNNN** [comment]
This RFC is just only 1 day old. Independently from the 1/3 of negative votes, give the community time to evaluate this RFC. When the current state does not change, then we will close this issue for an better overview.

**Comment by lfkeitel** [comment]
I feel a mutable HTTP request/response would've been the best way to go. It makes the most since to me as there's only ever one request and one response. It someone needs a copy (for testing or whatnot), then they can clone an object.

However, it is not our place to propose a competing standards recommendation. That will just cause confusion and separation. Plus, you can create a wrapper around a PSR-7 library to fake mutability if you really want it.

Yes, for mutability. No, for creating a new standard for it.

**Comment by TorbenKoehn** [comment]
@lfkeitel I don't know what you're developing in your free-time and at work, but there aren't many applications where I only have one request, since I usually to a shitload of other HTTP requests (e.g. accessing JSON APIs of any kind etc.)

Personally I think the Immutable-Interface was the only way to go, since you want to use them in middleware and that way you can always make sure, that a middleware can't modify the request you're working with right now.

Immutability also helps reducing bugs (primarily because of above reasons)

I'm absolutely against changing the current PSR-7 standard, it's really fine mostly.

The only annoying thing with PSR-7 is the fixed Stream-implementation in there, if I write e.g. a Stream- or FS-library, I always reference PSR-7 completely just to implement the Stream-interface.

**Comment by lfkeitel** [comment]
That was a wording issue on my part. I meant there's only ever one _client_ request and response at a time. By its nature PHP runs a script for each client. Each run has a canonical request and has a canonical response back to the client. The request in this case I can understand being immutable. The response however I feel should still be mutable.

I'm fully aware there are times where we need to make HTTP requests in the application itself in order to fulfill the request. In that case a mutable request makes sense because it is being built up but it still represents that one request. But the response in that case should be immutable since the application received it from an external source.

All I meant was that there are times where mutability makes sense. To me at least. Everyone seems to have their own opinions and that's fine.

Ultimately, the point of my response was to explain why I didn't feel we should create a competing recommendation. I didn't mean to get in a debate about mutability vs immutability.

**Comment by mathroc** [comment]
there are lot of use cases where there is more than one request targeting the app at the same time (same instance of your app/container). eg: [sub requests](http://symfony.com/doc/current/components/http_kernel/introduction.html#sub-requests) or when running the server in PHP (eg: https://github.com/php-pm/php-pm ) but none of that matters regarding mutability or immutability!

**Comment by TorbenKoehn** [comment]
@mathroc It matters a lot, really.

It's the difference between knowing if whatever runs behind `$next($request, $response)` will be able to modify `$request` and `$response` or not.

**Comment by mathroc** [comment]
@TorbenKoehn absolutely! mutability or immutability matters a lot, I agree completely!
 what does not is if there is one or more requests handled at the same time. it can be done both with mutable and immutable object, the solution will "just" be different

**Comment by sagikazarmark** [comment]
I think immutability was a quite good and wise decision of the PSR editors. HTTP Messages are Value Objects which means their value is what you can identify them by, not their object reference. -> immutability.

Also, with immutability you can use PSR-7 with a much more functional approach, which does not allow the shared state (this is why most people complain), but it's a good thing. Makes your application much more readable and maintainable.

I think it cannot be compared to Symfony at all. Immutability works quite well with middlewares (chain of responsibility), while Symfony's mutable implementation uses events for a similar behaviour. Symfony makes the whole thing quite convenient, but it does not make things better at all from an architecture POV. The usability argument is not proven to me, and in the end, it rather seems to be about convenience and not sane logic.

To sum up: I think the functional approach will always be superior to the mutable one. The FIG came to this conclusion after a very long discussion.
